### PR TITLE
Chore/perf retain values

### PR DESCRIPTION
**Pull request by jacinta-stacks**
Related to [native function rework](https://github.com/stx-labs/core-epics/issues/117)

I did change the retain_values such that the sequence_data may be corrupted during retain_values call if it fails but the only caller of this function immediately drops it so it doesn't make a difference. If its important to keep, I can add a two pass approach: 
1. First pass: evaluate all predicates, store results in a Vec<bool>
2. If any errors, return early — sequence untouched
3. Second pass: swap-to-front using the bitmap, then truncate

But given the caller drops on error anyway, it seemed like unnecessary added complexity but please scream if you rather I do that.

**Benchmarking Results**
Size | old (quadratic) | new (linear) | Speedup
-- | -- | -- | --
100×5 | 13.6 µs | 10.4 µs | 1.31×
500×5 | 122.9 µs | 50.8 µs | 2.42×
500×10 | 138.2 µs | 70.9 µs | 1.95×

### CODE REVIEW

**Comment by benjamin-stacks** [comment]
> but the only caller of this function immediately drops it so it doesn't make a difference

My concern with this is that `clarity-types` is a library crate that may very well be used by other projects which, unlike the internal reference you found, might rely on the behavior. It's also philosophically similar to transactions not committing if there's any error, which would be a familiar concept to people working with Clarity.

I would either rename the function (ensuring that any user of the function gets a breakage on upgrade and thus is forced to have a closer look), or retain (haha) the existing error behavior.

**Comment by aaronb-stacks** [comment]
> I would either rename the function (ensuring that any user of the function gets a breakage on upgrade and thus is forced to have a closer look), or retain (haha) the existing error behavior.

I think we may have a better path -- we can update this function to take ownership of `self` (we could rename it too) and then `Ok` returns `Ok(self)`. That way, there's no "option" to attempt to recover the original list in the error case, and the interface forces any caller who would have depended on this to do something else (e.g., write their own function, or complain on Github).

File: clarity-types/src/tests/types/mod.rs [hunk 1]
```diff
@@ -627,3 +659,240 @@ fn invalid_utf8_string_from_bytes() {
...
+    let result = seq.try_retain::<(), _>(utils::keep_none).unwrap();
+    assert_eq!(
+        SequenceData::String(CharType::UTF8(UTF8Data { data: vec![] })),
+        result
+    );
+}
+
+#[test]
+fn test_sequence_try_retain_utf8_multibyte_filter() {
+    let dog = vec![0xF0, 0x9F, 0x90, 0xB6]; // 🐶
```
Comment (cylewitruk-stacks) [inline clarity-types/src/tests/types/mod.rs:848]: 🤣

**Review by cylewitruk-stacks — APPROVED** [review]
lgtm

**Comment by federico-stacks** [comment]
> LGTM, but would you mind adding a test or two? This function has no direct unit tests (as evidenced by the fact that it could be renamed), which is fine, but the coverage via the Clarity `filter` tests isn't particularly exhaustive either.
> 
> Feels like with a non-trivial algorithm like this, the usual edge cases (everything is retained; nothing is retained; the sequence was empty to begin with; maybe something with the first/last elements specifically) should have coverage.

@benjamin-stacks test coverage addressed here: ee56ffd4c73759ed2d12424e40f433409adb5c47. This PR should be ready to go!

File: clarity-types/src/tests/types/mod.rs (hunk 1 above)
Comment (benjamin-stacks) [reply clarity-types/src/tests/types/mod.rs:848]: Nit: this should be a unicorn because unicorns are awesome

Comment (federico-stacks) [reply clarity-types/src/tests/types/mod.rs:848]: I also added a `cat` on the next line, we should be fine anyway :)
